### PR TITLE
Replace Boost PP_Iterate by C++11 variadic templates

### PR DESCRIPTION
**Pull request by Flamefire**
This is basically the continuation of #88

It allows support for (almost) any number of arguments and makes setting MOCK_MAX_ARGS and MOCK_MAX_SEQUENCES unnecessary. The only remaining limitation is imposed by Boost.Preprocessor in `BOOST_PP_REPEAT` which has a large limit (e.g. 256 in 1.69)
It also allows for easier debugging due to being able to step into actual code instead of preprocessor generated stuff

Not sure how well git recognizes this but I basically removed the suffix from the `*_template.hpp` files and replaced the `BOOST_PP_REPEAT/BOOST_PP_ENUM` macros by variadic templates and parameter pack expansions.

It also cuts down on the use of macros which should make the code a bit cleaner and easier to follow. Especially in regard to r-value-references and move-only types (`unique_ptr`...) I suspect a few issues. I'll also include a few cleanups with template aliases I noticed while doing this, e.g. `typename arg_type<...>::type`

### CODE REVIEW

File: test/test_max_args.cpp [hunk 1]
```diff
@@ -12,6 +12,8 @@
 #include <boost/test/unit_test.hpp>
 
 #define IDENTITY(z, n, d) d
+// Number of arguments is (now) unlimitted, so take any high value here
+#define MOCK_MAX_ARGS 100
```
Comment (mat007) [inline test/test_max_args.cpp:16]: This is now unneeded to define this, isn’t it?

**Review by mat007 — COMMENTED** [review]
Less preprocessor is always good 😁

File: test/test_max_args.cpp (hunk 1 above)
Comment (Flamefire) [reply test/test_max_args.cpp:16]: It is used below to generate functions using that many args. Will rename the define.

**Comment by Flamefire** [comment]
Rebased after #112, now only 7 commits with changed line stats of "+440 −3,924" :D

Ok, most of the removed lines are in the benchmark files which are no longer required, but still: Less code ;-)

**Comment by Flamefire** [comment]
@mat007 After this is merged, I'd prepare a release 2.0.0. The last release is from 19 June 2020, so it's about time... ;-)
And it needs to be a new major version due to the breaking changes. It might also serve as a proof that this great project is still alive!

I'd probably include fixes for the (small) issues #94, #81 and check https://github.com/mat007/turtle/issues/69#issuecomment-657057644 (might also be a small breaking change, not sure)

Support for method qualifiers (noexcept, rvalue, lvalue, ...) and override will need to wait for the next version or so.

I'd also like to rename the `MOCK_METHOD*_EXT` macros. They are basically the non-variadic form of `MOCK_METHOD*` and hence no longer required as the variadic form is basically exactly the same. I'd go for `MOCK_METHOD_IMPL`, for "implementation"   
Reason is, that for the qualifiers, I'd use that name and hence want to "free" it in this release, see the idea at https://github.com/mat007/turtle/issues/48#issuecomment-663835669 which could become something like:

    MOCK_METHOD(name, cardinality, opt. list of specifiers, opt. id) -> for each entry in the specifier list an overload is generated
    MOCK_METHOD_EXT(name, cardinality, signature, id, opt. list of specifiers) -> similar

Details don't really matter. Basically: These macros are no longer required and (now) undocumented similar to the TPL-macros, so I'd remove them while we're at it, so we can reuse the name later on for maybe something really different.

**Pull request merged**